sklearn/kernel_approximation.py

```diff
@@ -494,13 +494,19 @@ def fit(self, X, y=None):
             Returns the instance itself.
         """
         self._validate_params()
-        X = self._validate_data(X)
+        X = self._validate_data(X, dtype=[np.float64, np.float32])
         random_state = check_random_state(self.random_state)
         n_features = X.shape[1]
         uniform = random_state.uniform(size=(n_features, self.n_components))
         # transform by inverse CDF of sech
-        self.random_weights_ = 1.0 / np.pi * np.log(np.tan(np.pi / 2.0 * uniform))
-        self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components)
+        self.random_weights_ = (
+            1.0
+            / np.pi
+            * np.log(np.tan(np.pi / 2.0 * uniform)).astype(X.dtype, copy=False)
+        )
+        self.random_offset_ = random_state.uniform(
+            0, 2 * np.pi, size=self.n_components
+        ).astype(X.dtype, copy=False)
```

Comment:
As was decided in #24317, let's just convert self.random_weights_ and self.random_offset_ afterwards if X is float32. Sorry for my previous misleading comment, you can revert to your original implementation.